### DISCUSSION TITLE
Livewire not working with mix.browserSync

### DISCUSSION THREAD

**Discussion by natanaelsinisalo** [livewire/livewire#3127] [7. Bugs (archived)]
### Description

Okay, so I first don't know if Livewire is supposed to work with mix.browserSync, so please let me know if it's just plain not supported. I see that there have been issues #996 and #1130 touching this topic and in #996 it seems that the issue author got it working by updating Livewire to 1.2.0. I'm using Livewire version 2.3.15.

I tried to start using Livewire in our Laravel 8 application that is in development and followed the quick start guide to implement the example counter. I had this error:

_Livewire encountered corrupt data when trying to hydrate the [counter] component. Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests._ 

After some investigation I found out that Livewire works with the url https://projectname.valet.test but not with the Browsersync proxy url: https://localhost:3000

To make debugging easier I installed a fresh Laravel 8 instance in url https://laravel-8.valet.test with just livewire/livewire package, configured Browsersync like this: `mix.browserSync('https://laravel-8.valet.test');` and followed the counter functionality in the [quick start guide](https://laravel-livewire.com/docs/2.x/quickstart). That gave me the very same result:

 _Livewire encountered corrupt data when trying to hydrate the [counter] component. Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests._ 

So I assumed that Livewire checks the APP_URL and compares it with the request domain and changed APP_URL to match https://localhost:3000, but still had the same issue.

Is this a bug? Is there something I should do to get this working? Or is Livewire just not compatible with mix.browserSync?

### Exact steps to reproduce

1.) Install fresh laravel/laravel 8.5.9
2.) Install Livewire 2.3.15
3.) Add `mix.browserSync('https://yordomain.test');` into your webpack.mix.js
4.) Run npm i && npm run watch
5.) Start implementing the [counter functionality](https://laravel-livewire.com/docs/2.x/quickstart).

**Context**
- Livewire version: 2.3.15
- Laravel version: 8.5.9 (laravel/framework 8.26.1)
- Alpine version: none
- Browser: Firefox Developer Edition

**Comment by KillTomer** [comment]
The same error here for every Livewire component after the Livewire update v2.3.15.  Laravel Mix v6.0.11. PHP 8, Laravel 8.
It doesn't work only with browserSync().

**Comment by jleontidis** [comment]
Same for me. I did some downgrading and testing and found out that version v2.3.8 does not have this issue.

**Comment by roni-estein** [comment]
so terrible. this makes me sad, although I thought I was going insane for about 2 hours so at least that's done!

**Comment by halibut-jake** [comment]
https://github.com/livewire/livewire/commit/2ff70fdd893ecf9860fb70105efbc81aee93a02a

the effect of changing `Livewire::originalUrl()`  to return `url()->current()`


### Temporary solution (v2.3.16)
```javascript
// webpack.mix.js
mix.browserSync({
  proxy: 'localhost:8000',
  open: false,
});
```

```blade
<!-- layout.blade -->
<html>
  ...
  <body>
    ...
    @if (app()->isLocal())
      <script src="http://localhost:3000/browser-sync/browser-sync-client.js"></script>
    @endif
  </body>
</html
```

Run and open http://localhost:8000
Now Livewire with browserSync works fine in APP_ENV=local

**Comment by roni-estein** [comment]
oh wow! you just need to configure browsersync to run on port 8000 instead of port 3000?

**Comment by jleontidis** [comment]
I tried @halibut-jake solution but it is not working for me.

What is working is changing **Livewire::originalUrl()** with an empty string in LifecycleManager.php file. 
**It is not a solution rather a workaround until a real solution is found.**

`public static function fromInitialRequest($name, $id)
    {
        return tap(new static, function ($instance) use ($name, $id) {
            $instance->instance = app('livewire')->getInstance($name, $id);
            $instance->request = new Request([
                'fingerprint' => ['id' => $id, 'name' => $name, 'locale' => app()->getLocale(), 'url' => ''],
                'updates' => [],
                'serverMemo' => [],
            ]);
        });
    }`

**Comment by roni-estein** [comment]
@halibut-jake thanks, that didn't work for me either. @jleontidis can you further explain what is happening in that code? a little after midnight and I was just wondering if we are introducing some security risk to the component?

**Comment by halibut-jake** [comment]
Browsersync continues to run on 3000 ports.
Even if you don't access to 3000 ports, you'll still be able to get browserync.js.
I checked that it works in v2.3.16.

**Comment by marijoo** [comment]
Same here.

**Comment by roni-estein** [comment]
I've tried just about every permutation of this, and I've gotten it to work, but there are always weird side effects from livewire itself not working properly to others. If someone finds a clear non-hack methodology to work with browser-sync, please post it here. I'd be happy to move forward, as it is it's robbed too many hours from my work week.

I will say not just on the browser-sync front, but it is very frustrating and disheartening to have so many core changes happening in minor patch releases. This whole issue could have come about by releasing this as 2.4.0. Now I do get that this is coming from patching other items that had issues as well, and there are always tradeoffs, but this isn't the first breaking change released as a minor patch. As I use Livewire in a lot of production work, I am really kind of nervous about delegating any upgrades at this point. I love and support livewire both with money and time on discord. Please take some time to think about semantic versioning and major-minor releases. I'm fine with things having issues, but blowing me up in a safe composer update shouldn't happen. For now, I'm just locking all my projects on. 2.3.8.

**Comment by basepack** [comment]
Same, downgrading to Livewire 2.3.8 fixed it. For now I just locked Livewire on version 2.3.8 via composer.json.

I wholeheartedly agree what @roni-estein pointed out in his last paragraph.

**Comment by calebporzio** [comment]
@roni-estein, completely agree. However, none of the breaking changes were intended, and all are trying to be addressed.

This is all coming about because of a necessary security patch. I'm trying literally my best with this. Please understand.

Agreed, I should have tagged 2.4 for that fix, but from a semvar perspective it really is a patch in the sense that it plugs a hole. But it did an a new API and like people have mentioned broke some things, so in hindsight it should have been 2.4.

Anyhow, sorry for the trouble, I really am. But all of this had to be done. After this pain, Livewire will go back to stability. This is an exceptional case and not the way Livewire will behave or ever has behaved. We are dedicated to stability and compatibility around here.

**Comment by calebporzio** [comment]
(also working on a fix for this)

**Comment by marijoo** [comment]
As much as I understand the displeasure – isn’t this really just a bug about using BrowserSync? Sure, it is a bit uncomfortable to go without BrowserSync if you are used to it, but the function of Livewire itself is still given, isn’t it?

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by calebporzio** [comment]
Fixed and tagged: v2.3.17

Re-open if I'm wrong.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by roni-estein** [comment]
> As much as I understand the displeasure – isn’t this really just a bug about using BrowserSync? Sure, it is a bit uncomfortable to go without BrowserSync if you are used to it, but the function of Livewire itself is still given, isn’t it?

I’m only referencing browser-sync here because that’s the start of this issue. But no. You should source dive the changes from 2.3.9 on, they are extensive. This is one.

**Comment by roni-estein** [comment]
> Fixed and tagged: v2.3.17
> 
> Re-open if I'm wrong.

Checking in 5 thanks!

**Comment by roni-estein** [comment]
@calebporzio Nice!!! 
3.17 mostly working! However, file uploads still won't work under browser-sync. While everything is hot and fresh in our heads think we can bring this home?

**Comment by basepack** [comment]
Still have this error in some instances where I show forms, **without** file uploads.

**Comment by danieltorscho** [comment]
Same here, still not working

**Comment by calebporzio** [comment]
@danieltorscho , have you dig into the problem at all? Can you offer up more information?

**Comment by roni-estein** [comment]
I've been DEEP into laravel migration land for a while but I'm back in livewire action. I can check it out this tonight after I get the kids to bed, and see if file uploads work in the current browsersync pretty quickly. I'll run the redirect intended too, and give you an update on both and see if I can find anything.

**Reply by basepack** [reply]
Hi @roni-estein did you find anything? This browsersync issue still persist since a long time.

**Reply by roni-estein** [reply]
Oh No! 
I got so swamped and I turned off all but critical notifications without dealing with this! sorry! I'll take the blame 😬. No File uploads and some redirects possibly other things still don't work in Livewire running BrowserSync as of version 2.7.2 :( 

With any luck we have @joshhanley going deep on this bad boy! 🎉 If you need anything Josh, I'll stop on a dime to help, test, or even just to  get you coffee... whatever it takes to get this working :)

**Reply by basepack** [reply]
Haha! No blame taken roni-estein. For now I am hitting F5 where that hydrate error pops up.

**Comment by joshhanley** [comment]
@roni-estein and @basepack 

Mmm coffee...

Oh sorry, right the issue, I've dug into this before with Roni, and the issue is browsersync is actually proxying the connection (if I remember rightly), hence why when Livewire tries to get the app url, it's actually wrong (or something like that).

One way you may be able to get this running is to do this:

In your app layout file, add this right at the end before the closing body tag (change the url to your app url but make sure to keep the `:3000`)

```blade
@if (app()->isLocal())
    <script src="http://livewireplayground.test:3000/browser-sync/browser-sync-client.js"></script>
@endif
```

Then in your `webpack.mix.js` file, change your browsersync configuration to this (again change url but this time without the `:3000`)

```js
.browserSync({
    proxy: 'livewireplayground.test',
    // host: 'livewireplayground.test', // you shouldn't need this option, but if it's not playing happily you could uncomment and use this too
    open: false, // This just stops browsersync automatically opening a browser window as it's not needed, see below
})
```

Then when you run `npm run watch`, and browse to your normal app url, you should see the "browsersync: connected" appear in the top right.

This means it's using the correct url for everything (rather than proxying your app through localhost:3000), so in theory the file uploads should work.

I did a quick test and I could see the temporary uploaded file was working (but I didn't test further than that).

Test that out and let me know how it goes!

Hope this helps!

**Comment by roni-estein** [comment]
Josh you amazing NINJA!!! if only I had checked my email before going to bed… ok I’m getting up, even though I have to then take my excited puppy out too … in the snow … in my underwear and t shirt but dammit! This is worth it!
  
  
  
    
  

  
  
>

**Reply by roni-estein** [reply]
![Screen Shot 2021-10-26 at 1 06 14 AM](https://user-images.githubusercontent.com/8517475/138818454-f286ff90-1482-42f6-bd6b-e29b17750d75.png)
At a quick Glance I wasn't able to make it work, I'll look at it in the morning with some fresh eyes and go through all the browsersync configs again. I did notice this hanging around from browser sync so perhaps onceI'm able to eliminate the second one it may work...

![Screen Shot 2021-10-26 at 1 09 19 AM](https://user-images.githubusercontent.com/8517475/138818658-1dec5438-e3a9-48cd-9ad7-22c40bef19a8.png)

**Reply by roni-estein** [reply]
That's with all the steps followed
![Screen Shot 2021-10-26 at 1 10 12 AM](https://user-images.githubusercontent.com/8517475/138818814-5937c730-e2af-409c-9c1c-1b19ae64a82e.png)
..

**Reply by joshhanley** [reply]
@roni-estein haha! How'd you go?

**Comment by jeffegiovani** [comment]
The same error here for Livewire 2.8.2

**Reply by joshhanley** [reply]
@jeffegiovani when you say same error, do you mean corruption error? Did you see my explanation above as to why they don't work out of the box and try my suggestion?
